src/services/webhdfs/backend.rs

```diff
@@ -0,0 +1,1026 @@
+            Err(e) => match e.kind() {
+                ErrorKind::ObjectNotFound => {
+                    debug!("working directory does not exist, creating...");
+                    backend.blocking_create("/", OpCreate::new(ObjectMode::DIR))?;
+                }
+                _ => return Err(e),
+            },
+        }
+        debug!("working directory is ready.");
+        Ok(backend)
+    }
+}
+
+/// Backend for WebHDFS service
+#[derive(Debug, Clone)]
+pub struct Backend {
+    root: String,
+    endpoint: String,
+    client: HttpClient,
+    auth: Option<String>,
+    csrf: String,
+}
+
+impl Backend {
+    async fn create_or_write_object_req(
+        &self,
+        path: &str,
+        size: Option<u64>,
+        content_type: Option<&str>,
+        body: AsyncBody,
+    ) -> Result<AsyncReq> {
+        let p = build_abs_path(&self.root, path);
+        let op = if path.ends_with('/') {
+            "MKDIRS"
+        } else {
+            "CREATE"
+        };
+        let mut url = format!(
+            "{}/webhdfs/v1/{}?op={}&overwrite=true&noredirect=true",
+            self.endpoint,
+            percent_encode_path(&p),
+            op,
+        );
+        if let Some(auth) = &self.auth {
+            url = url + "&" + auth;
+        }
+
+        let req = Request::put(&url)
+            .header(self.csrf.as_str(), "")
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+        // mkdir does not redirect
+        if path.ends_with('/') {
+            return Ok(req);
+        }
+
+        let resp = self.client.send_async(req).await?;
+
+        self.put_along_redirect(resp, size, content_type, body)
+            .await
+    }
+
+    async fn open_object_req(&self, path: &str, range: &BytesRange) -> Result<AsyncReq> {
+        let p = build_abs_path(&self.root, path);
+        let mut url = format!(
+            "{}/webhdfs/v1/{}?op=OPEN&noredirect=true",
+            self.endpoint,
+            percent_encode_path(&p),
+        );
+        if let Some(auth) = &self.auth {
+            url = url + "&" + auth;
+        }
+        let (offset, size) = (range.offset(), range.size());
+
+        match (offset, size) {
+            (Some(offset), Some(size)) => {
+                url = format!("{url}&offset={offset}&length={size}");
+            }
+            (Some(offset), None) => {
+                url = format!("{url}&offset={offset}");
+            }
+            (None, None) => {
+                // read all, do nothing
+            }
+            (None, Some(_)) => {
+                // handled outside of this function, should never happen
+                unreachable!()
+            }
+        }
+
+        let req = Request::get(&url)
+            .header(self.csrf.as_str(), "")
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        Ok(req)
+    }
+
+    fn status_object_req<B>(&self, path: &str, body: B) -> Result<Request<B>> {
```

Comment:
I prefer the API naming style with the underlying API name. For example: `webhdfs_get_file_status`.